### PR TITLE
Fix onboarding resume colliding with name/salutation sub-steps

### LINKED ISSUE #907: Onboarding: resume nudge collides with name/salutation sub-steps ("two flows at once")
## Symptom

During V2 onboarding (first-5-minutes WOW), a new user sees **two flows fire at once**:

1. When asked for their name and they type e.g. `P`, two messages arrive simultaneously — the Twin-onboarding **resume nudge** ("Bé Tiền đang chờ bạn ở bước *câu hỏi mục tiêu* … Tiếp tục nhé?") *and* the live first-5-min-wow salutation question.
2. Tapping the salutation/gender button again triggers two flows → two messages → and the live Reading/number-input flow can no longer be completed.

Net effect: the minute-1 intro (name → salutation → goal) gets forked and the user is stranded.

## Root cause

Name entry and salutation pick are **collapsed into the single `goal_question` DB step** (`OnboardingSession.current_step`); the live sub-step is only derivable from the `User` row. Crucially, those sub-steps mutate the **`User`** row (`set_display_name` / `set_salutation`), never the **`OnboardingSession`** row — so `OnboardingSession.updated_at` stays frozen at `/start` time while the user is on name/salutation.

Consequences:
- The resume-nudge cron (every 5 min; fires for sessions "stuck >10 min" on `goal_question`, one nudge per user ever) mis-classifies a name-entry user as *stuck at goal pick* and fires a Twin-centric nudge mid-intro — off-message and premature ("WOW first, Twin payoff").
- That nudge's **Tiếp tục** → `_resume_at(goal_question)` → `_send_goal_question()` **skips name + salutation**, forking the flow.

## Fix

No migration — derive the sub-step from existing columns:
- `onboarding_service.goal_substep(user)` → `name` → `salutation` → `goal`.
- `_resume_at` for `goal_question` resumes at the **correct sub-step** (re-send name prompt / salutation question / goal question) instead of jumping to goal.
- Resume-nudge job **skips** the name/salutation intro sub-steps (`_is_intro_substep`), leaving `nudge_sent_at` NULL so the user stays eligible once they reach goal pick / asset / twin.

Twin payoff at minute-4 is untouched (no regression).

## Acceptance
- [ ] No nudge fires during name/salutation sub-steps.
- [ ] Resume never skips name/salutation.
- [ ] Unit tests for `goal_substep`, `_resume_at` routing, and the job skip predicate.

### PR DESCRIPTION
**Pull request by phuongphh**
Close #907

## Problem

During V2 onboarding a new user saw **two flows fire at once**: typing their name (e.g. `P`) produced both the Twin-onboarding **resume nudge** ("Bé Tiền đang chờ bạn ở bước *câu hỏi mục tiêu* … Tiếp tục nhé?") *and* the live first-5-min-wow salutation question. Tapping the salutation button re-triggered the collision and stranded the Reading/number flow.

## Root cause

Name entry and salutation pick are collapsed into the single `goal_question` DB step — the live sub-step is only derivable from the `User` row. Those sub-steps mutate the **`User`** row (`set_display_name` / `set_salutation`), never the **`OnboardingSession`** row, so `OnboardingSession.updated_at` freezes at `/start` time.

So a name-entry user trips the resume-nudge cron's "stuck >10 min on goal_question" filter within minutes of starting. The cron then:
1. fires a Twin-centric nudge mid-intro (off-message; violates "WOW first, Twin payoff"), and
2. its **Tiếp tục** → `_resume_at(goal_question)` → `_send_goal_question()` **skips name + salutation**, forking the flow.

## Fix (no migration — sub-step derived from existing columns)

- **`onboarding_service.goal_substep(user)`** — pure helper deriving `name` → `salutation` → `goal` from `display_name` / `salutation`.
- **`_resume_at`** for `goal_question` now resumes at the *correct* sub-step (re-send name prompt / salutation question / goal question) instead of jumping straight to the goal question.
- **Resume-nudge job** skips the name/salutation intro sub-steps via `_is_intro_substep`, leaving `nudge_sent_at` NULL so the user stays eligible once they advance to goal pick / first asset / twin.

The Twin payoff at minute-4 is untouched — no onboarding-v2 regression.

## Tests

New `tests/test_phase_4_4/test_onboarding_resume_substep.py` (18 cases):
- `goal_substep` derivation incl. whitespace/None edge cases
- `_resume_at` routes to the right sub-step and never skips name/salutation
- job skip predicate true on name/salutation, false at goal pick and later steps

```
tests/test_phase_4_4/test_onboarding_resume_substep.py … 18 passed
tests/test_phase_4_4/ tests/test_phase_4_1/test_epic_a.py tests/test_phase_4_2/ … 190 passed
```

Ruff clean on all changed/new files (pre-existing unrelated lint in `onboarding_v2.py` untouched).

## Layer contract
- `goal_substep` is pure (no DB / env). All mutators remain flush-only; the job owns its commit. No new `telegram_service` imports in services.

### CODE REVIEW

File: backend/jobs/onboarding_resume_job.py [hunk 1]
```diff
@@ -141,6 +163,10 @@ async def run_onboarding_resume_job() -> int:
                 user = await db.get(User, session.user_id)
                 if not user:
                     continue
+                # WOW-first: don't fire the Twin-centric nudge during the
+                # name/salutation intro sub-steps (see _is_intro_substep).
+                if _is_intro_substep(session, user):
+                    continue
```
[the remark this answers is not in the document]
Comment (phuongphh) [reply backend/jobs/onboarding_resume_job.py:169]: Good catch — fixed in d3677a5. Added a flush-only `touch_session` mutator and call it when the user advances through the name → salutation → goal sub-steps, so the session's activity timestamp tracks the latest interaction instead of `/start`. This closes the gap where a user lingering in the intro would get nudged the instant they reached goal pick (the cron still saw the stale timestamp). Since name/salutation live on the `User` row, the session's `onupdate` never fired — `touch_session` sets `updated_at` explicitly to force the UPDATE. Covered by two new tests (`test_touch_session_bumps_updated_at`, `test_touch_session_missing_returns_none`).

---
_Generated by [Claude Code](https://claude.ai/code)_
